[Blazor] Templated components - TemplatedNavBar + keyed TableTemplate (#32360)
dotnet/AspNetCore.Docs: aspnetcore/blazor/components/templated-components.md

> 

Reviewer: On the preceding line, let's drop "couple" ... and it won't let me make a formal GH suggestion, so I'll place the last sentence in a DIFF ...

```diff
- A couple of examples include:
+ Examples include:
```

* A table component that allows a user to specify templates for the table's header, rows, and footer.
* A list component that allows a user to specify a template for rendering items in a list.

> * A nav-bar component that allows a user to specify a template for start content, and items.

Reviewer: ```suggestion
* A navigation bar component that allows a user to specify a template for start content and navigation links.
```

Reviewer: I'm a little concerned about the "start content" language here ... that it isn't immediately clear what it mean. However, it's easy to see in the sample code what it means. Therefore, let's go with it. If someone has a problem with it, they can open an issue saying that they're confused by it.

A templated component is defined by specifying one or more component parameters of type <xref:Microsoft.AspNetCore.Components.RenderFragment> or <xref:Microsoft.AspNetCore.Components.RenderFragment%601>. A render fragment represents a segment of UI to render. <xref:Microsoft.AspNetCore.Components.RenderFragment%601> takes a type parameter that can be specified when the render fragment is invoked.

> [!NOTE]
> For more information on <xref:Microsoft.AspNetCore.Components.RenderFragment>, see <xref:blazor/components/index#child-content-render-fragments>.

> Often, templated components are generically typed, as the following `TemplatedNavBar` component demonstrates. The generic type `<T>` in this example is used to render `IReadOnlyList<T>` values, which in this case is a list of pets for a component that displays a nav-bar with links to them.

Reviewer: ```suggestion
Often, templated components are generically typed, as the following `TemplatedNavBar` component demonstrates. The generic type (`<T>`) in the following example is used to render <xref:System.Collections.Generic.IReadOnlyList%601> values, which in this case is a list of pets for a component that displays a navigation bar with links to a pet detail component.
```

> `TableTemplate.razor`:

Reviewer: ```suggestion
`TemplatedNavBar.razor`:
```

> Often, templated components are generically typed, as the following `TemplatedNavBar` component demonstrates. The generic type (`<T>`) in the following example is used to render <xref:System.Collections.Generic.IReadOnlyList%601> values, which in this case is a list of pets for a component that displays a navigation bar with links to a pet detail component.

Reviewer: I missed one ...

```suggestion
Often, templated components are generically typed, as the following `TemplatedNavBar` component (`TemplatedNavBar.razor`) demonstrates. The generic type (`<T>`) in the following example is used to render <xref:System.Collections.Generic.IReadOnlyList%601> values, which in this case is a list of pets for a component that displays a navigation bar with links to a pet detail component.
```

:::moniker-end

> Alternatively, you can change the parameter name using the `Context` attribute on the <xref:Microsoft.AspNetCore.Components.RenderFragment%601> child element. In the following example, the `Context` is set on `ItemTemplate` rather than `TemplatedNavBar`:

Reviewer: And here ...

```suggestion
Alternatively, you can change the parameter name using the `Context` attribute on the <xref:Microsoft.AspNetCore.Components.RenderFragment%601> child element. In the following example, the `Context` is set on `ItemTemplate` rather than `TemplatedNavBar`.
```

:::moniker-end

> The example provided in `TemplatedNavBar.razor` assumes that the `Items` collection does not change after the initial render, or that if it does change, maintaining the state of components/elements used in `ItemTemplate` is not necessary. For templated components where such usage cannot be anticipated, see the following section.

Reviewer: ```suggestion
The example provided in the `TemplatedNavBar` component (`TemplatedNavBar.razor`) assumes that the `Items` collection doesn't change after the initial render; or that if it does change, maintaining the state of components/elements used in `ItemTemplate` isn't necessary. For templated components where such usage can't be anticipated, see the [Preserve relationships with `@key`]#preserve-relationships-with-key) section.
```

## Preserve relationships with `@key`

> Templated components are often used to render collections of items, such as tables or lists. In such general scenarios, we cannot assume that the user will not use stateful components/elements in the item template definition, or that there won't be additional changes to the `Items` data source collection. For such templated components, it is necessary to preserve the relationships with `@key` directive attribute.

Reviewer: ```suggestion
Templated components are often used to render collections of items, such as tables or lists. In such general scenarios, we can't assume that the user will avoid stateful components/elements in the item template definition or that there won't be additional changes to the `Items` collection. For such templated components, it's necessary to preserve the relationships with the `@key` directive attribute.
```

> [!NOTE]

> > For more information on `@key` directive attribute, see <xref:blazor/components/element-component-model-relationships>.

Reviewer: ```suggestion
> For more information on the `@key` directive attribute, see <xref:blazor/components/key>.
```

> The following `TableTemplate.razor` demonstrates a templated component where releationship are preserved with `@key`:

Reviewer: ```suggestion
The following `TableTemplate` component (`TableTemplate.razor`) demonstrates a templated component that preserves relationships with `@key`:
```

> The following `TableTemplate` component (`TableTemplate.razor`) demonstrates a templated component that preserves relationships with `@key`:

Reviewer: Need a period here ... I missed this, too! ........ It's ***Friday***. My 🧠 is a bit 🔥 for the week!

```suggestion
The following `TableTemplate` component (`TableTemplate.razor`) demonstrates a templated component that preserves relationships with `@key`.
```

`TableTemplate.razor`:

:::moniker range=">= aspnetcore-8.0"

:::code language="razor" source="~/../blazor-samples/8.0/BlazorSample_BlazorWebApp/Components/TableTemplate.razor" highlight="10":::

:::moniker-end

:::moniker range=">= aspnetcore-7.0 < aspnetcore-8.0"

:::code language="razor" source="~/../blazor-samples/7.0/BlazorSample_WebAssembly/Shared/templated-components/TableTemplate.razor" highlight="10":::

:::moniker-end

:::moniker range=">= aspnetcore-6.0 < aspnetcore-7.0"

:::code language="razor" source="~/../blazor-samples/6.0/BlazorSample_WebAssembly/Shared/templated-components/TableTemplate.razor" highlight="10":::

:::moniker-end

:::moniker range=">= aspnetcore-5.0 < aspnetcore-6.0"

:::code language="razor" source="~/../blazor-samples/5.0/BlazorSample_WebAssembly/Shared/templated-components/TableTemplate.razor" highlight="10":::

:::moniker-end

:::moniker range="< aspnetcore-5.0"

:::code language="razor" source="~/../blazor-samples/3.1/BlazorSample_WebAssembly/Shared/templated-components/TableTemplate.razor" highlight="10":::

:::moniker-end

> Consider the `Pets5.razor` example below, which demonstrates the issue. In the component, each iteration of adding a pet in `OnAfterRenderAsync` results in Blazor rerendering the `TableTemplate.razor`.

Reviewer: ```suggestion
Consider the following `Pets5` component (`Pets5.razor`), which demonstrates the importance of keying data to preserve model relationships. In the component, each iteration of adding a pet in [`OnAfterRenderAsync`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync) results in Blazor rerendering the `TableTemplate` component.
```

`Pets5.razor`:

:::moniker range=">= aspnetcore-8.0"

:::code language="razor" source="~/../blazor-samples/8.0/BlazorSample_BlazorWebApp/Components/Pages/Pets5.razor":::

:::moniker-end

:::moniker range=">= aspnetcore-7.0 < aspnetcore-8.0"

:::code language="razor" source="~/../blazor-samples/7.0/BlazorSample_WebAssembly/Pages/templated-components/Pets5.razor":::

:::moniker-end

:::moniker range=">= aspnetcore-6.0 < aspnetcore-7.0"

:::code language="razor" source="~/../blazor-samples/6.0/BlazorSample_WebAssembly/Pages/templated-components/Pets5.razor":::

:::moniker-end

:::moniker range=">= aspnetcore-5.0 < aspnetcore-6.0"

:::code language="razor" source="~/../blazor-samples/5.0/BlazorSample_WebAssembly/Pages/templated-components/Pets5.razor":::

:::moniker-end

:::moniker range="< aspnetcore-5.0"

:::code language="razor" source="~/../blazor-samples/3.1/BlazorSample_WebAssembly/Pages/templated-components/Pets5.razor":::

:::moniker-end

This demonstration allows you to:

* Select an `<input>` from among several rendered table rows.
* Study the behavior of the page's focus as the pets collection automatically grows.

> Without using the `@key` directive attribute in `TableTemplate.razor`, the page's focus remains on the same index position of table row, causing the focus to shift each time a pet is added.

Reviewer: ```suggestion
Without using the `@key` directive attribute in the `TableTemplate` component, the page's focus remains on the same index position (row) of the table, causing the focus to shift each time a pet is added. To demonstrate this, remove the `@key` directive attribute and value, restart the app, attempt to modify a field value as items are added.
```

## Additional resources

* <xref:blazor/performance#define-reusable-renderfragments-in-code>

> * <xref:blazor/components/element-component-model-relationships>

Reviewer: ```suggestion
* <xref:blazor/components/key>
```
